chips/ibex/src/aes.rs

```diff
@@ -0,0 +1,347 @@
+                        _ => {}
+                    }
+                }
+                true
+            },
+        );
+        true
+    }
+
+    fn set_key(&self, key: &[u8]) -> ReturnCode {
+        let regs: &AesRegisters = &*self.registers;
+
+        loop {
+            if self.idle() {
+                break;
+            }
+        }
+
+        if key.len() != AES128_KEY_SIZE {
+            return ReturnCode::EINVAL;
+        }
+
+        for i in 0..4 {
+            let mut k = key[i * 4 + 0] as usize;
+            k |= (key[i * 4 + 1] as usize) << 8;
+            k |= (key[i * 4 + 2] as usize) << 16;
+            k |= (key[i * 4 + 3] as usize) << 24;
+            match i {
+                0 => regs.key0.set(k as u32),
+                1 => regs.key1.set(k as u32),
+                2 => regs.key2.set(k as u32),
+                3 => regs.key3.set(k as u32),
+                _ => {}
+            }
+        }
+
+        // We must write the rest of the registers as well
+        regs.key4.set(0);
+        regs.key5.set(0);
+        regs.key6.set(0);
+        regs.key7.set(0);
+        ReturnCode::SUCCESS
+    }
+
+    fn crypt(&self, start_index: usize, stop_index: usize, wr_start_index: usize) -> bool {
+        // convert our start_indexes in the array to block numbers
+        let start_block = start_index / AES128_BLOCK_SIZE;
+        let end_block = stop_index / AES128_BLOCK_SIZE;
+        let mut wr_block = wr_start_index / AES128_BLOCK_SIZE;
+        for i in start_block..end_block {
+            self.write_block(wr_block);
+            self.trigger();
+            self.read_block(i);
+            wr_block = wr_block + 1;
+        }
+        return true;
+    }
+}
+
+impl hil::symmetric_encryption::AES128<'a> for Aes<'a> {
+    fn enable(&self) {
+        self.configure(true);
+    }
+
+    fn disable(&self) {
+        self.clear();
+    }
+
+    fn set_client(&'a self, client: &'a dyn symmetric_encryption::Client<'a>) {
+        self.client.set(client);
+    }
+
+    fn set_iv(&self, _iv: &[u8]) -> ReturnCode {
+        // nothing because this is ECB
+        ReturnCode::SUCCESS
+    }
+
+    fn start_message(&self) {
+        return;
+    }
+
+    fn set_key(&self, key: &[u8]) -> ReturnCode {
+        return self.set_key(key);
+    }
+
+    fn crypt(
+        &'a self,
+        source: Option<&'a mut [u8]>,
+        dest: &'a mut [u8],
+        start_index: usize,
+        stop_index: usize,
+    ) -> Option<(ReturnCode, Option<&'a mut [u8]>, &'a mut [u8])> {
+        if stop_index - start_index > MAX_LENGTH {
+            return Some((ReturnCode::EINVAL, source, dest));
+        }
+        self.dest.replace(dest);
+        match source {
+            None => {
+                self.crypt(start_index, stop_index, start_index);
+            }
+            Some(src) => {
+                self.source.replace(src);
+                self.crypt(start_index, stop_index, 0);
+            }
```

Comment:
Do you think there would be a different way of expressing the "normal" vs "source = dest" cases?  It seems like something easy to misinterpret to the casual eye.

Reply:
Yeah, it's really confusing. The API really doesn't make this obvious what's going on. I think this is probably best fixed with a comment.